### PR TITLE
feat: improve the layout of secrets into their components

### PR DESCRIPTION
**Pull request by cardoe**
Moved all most of the secrets around into their respective components so that they are applied by the AppSet with the Application (component) that they are part of. Fixed where the mariadb secret was never being applied after converting to the AppSet. Ensure we always template out the OpenStack Helm secrets to make sure the data is always in sync.

### CODE REVIEW

**Review by nicholaskuechler — APPROVED** [review]

File: scripts/gitops-secrets-gen.sh [hunk 1]
```diff
@@ -413,20 +148,240 @@ spec:
...
+
+    # environment variable names
+    RABBITMQ_PASSWORD="$(echo "${component}_RABBITMQ_PASSWORD" | tr '[:lower:]' '[:upper:]')"
+    DB_PASSWORD="$(echo "${component}_DB_PASSWORD" | tr '[:lower:]' '[:upper:]')"
+    KEYSTONE_PASSWORD="$(echo "${keystone_user}_KEYSTONE_PASSWORD" | tr '[:lower:]' '[:upper:]')"
+
+    # k8s secret names
+    SECRET_RABBITMQ_PASSWORD="$(echo "${RABBITMQ_PASSWORD}" | tr '[:upper:]' '[:lower:]' | tr '_' '-')"
+    SECRET_DB_PASSWORD="$(echo "${DB_PASSWORD}" | tr '[:upper:]' '[:lower:]' | tr '_' '-')"
+    SECRET_KEYSTONE_PASSWORD="$(echo "${KEYSTONE_PASSWORD}" | tr '[:upper:]' '[:lower:]' | tr '_' '-')"
```
Comment (skrobul) [inline scripts/gitops-secrets-gen.sh:266]:
```suggestion
convert_to_secret_name() {
    echo "$1" | tr '[:upper:]' '[:lower:]' | tr '_' '-'
}

SECRET_RABBITMQ_PASSWORD=$(convert_to_secret_name "${RABBITMQ_PASSWORD}")
SECRET_DB_PASSWORD=$(convert_to_secret_name "${DB_PASSWORD}")
SECRET_KEYSTONE_PASSWORD=$(convert_to_secret_name "${KEYSTONE_PASSWORD}")
```

File: scripts/gitops-secrets-gen.sh [hunk 2]
```diff
@@ -413,20 +148,240 @@ spec:
...
+        echo "Writing ${component}/secret-keystone-password.yaml, please commit"
+        kubectl --namespace openstack \
+        create secret generic "${SECRET_KEYSTONE_PASSWORD}" \
+        --type Opaque \
+        --from-literal=username="${keystone_user}" \
+        --from-literal=password="${!KEYSTONE_PASSWORD}" \
+        --dry-run=client -o yaml \
+        | secret-seal-stdin "${DEST_DIR}/${component}/secret-keystone-password.yaml"
+    fi
+done
```
Comment (skrobul) [inline scripts/gitops-secrets-gen.sh:331]:
```suggestion

create_os_secret() {
    local file_suffix=$1
    local secret_name=$2
    local username=$3

    local component_var="REPLACE_${file_suffix}"

    if [ "x${!component_var}" = "xyes" ]; then
        echo "Writing ${component}/secret-${file_suffix,,}.yaml, please commit"
        kubectl --namespace openstack \
        create secret generic "${secret_name}" \
        --type Opaque \
        --from-literal=username="${username}" \
        --from-literal=password="${!file_suffix}" \
        --dry-run=client -o yaml \
        | secret-seal-stdin "${DEST_DIR}/${component}/secret-${file_suffix,,}.yaml"
    fi
}

create_os_secret "RABBITMQ_PASSWORD" "${SECRET_RABBITMQ_PASSWORD}" "${component}"
create_os_secret "DB_PASSWORD" "${SECRET_DB_PASSWORD}" "${component}"
create_os_secret "KEYSTONE_PASSWORD" "${SECRET_KEYSTONE_PASSWORD}" "${keystone_user}"

```

File: scripts/gitops-secrets-gen.sh [hunk 3]
```diff
@@ -413,20 +148,240 @@ spec:
...
+    echo "Checking ${component}"
+    mkdir -p "${DEST_DIR}/${component}/"
+    # keystone service account username
+    [ "x${component}" = "xkeystone" ] && keystone_user="admin" || keystone_user="${component}"
+    KEYSTONE_USER=$(echo "${keystone_user}" | tr '[:lower:]' '[:upper:]')
+
+    # environment variable names
+    RABBITMQ_PASSWORD="$(echo "${component}_RABBITMQ_PASSWORD" | tr '[:lower:]' '[:upper:]')"
+    DB_PASSWORD="$(echo "${component}_DB_PASSWORD" | tr '[:lower:]' '[:upper:]')"
+    KEYSTONE_PASSWORD="$(echo "${keystone_user}_KEYSTONE_PASSWORD" | tr '[:lower:]' '[:upper:]')"
```
Comment (skrobul) [inline scripts/gitops-secrets-gen.sh:261]:
```suggestion
    RABBITMQ_PASSWORD="$(echo "${component^^}_RABBITMQ_PASSWORD")"
    DB_PASSWORD="$(echo "${component^^}_DB_PASSWORD"')"
    KEYSTONE_PASSWORD="$(echo "${keystone_user^^}_KEYSTONE_PASSWORD")"
```

File: scripts/gitops-secrets-gen.sh [hunk 4]
```diff
@@ -413,20 +148,240 @@ spec:
...
+    create secret generic mariadb \
+    --dry-run=client \
+    -o yaml \
+    --type Opaque \
+    --from-literal=root-password="$(./scripts/pwgen.sh)" \
+    --from-literal=password="$(./scripts/pwgen.sh)" \
+    | secret-seal-stdin "${DEST_DIR}/openstack/secret-mariadb.yaml"
+
+# create constant OpenStack memcache key to avoid cache invalidation on deploy
+MEMCACHE_SECRET_KEY=$(cat "${DEST_DIR}/secret-openstack.yaml" 2>/dev/null | yq '.endpoints.oslo_cache.auth.memcache_secret_key')
```
Comment (skrobul) [inline scripts/gitops-secrets-gen.sh:238]:
```suggestion
MEMCACHE_SECRET_KEY=$(yq '.endpoints.oslo_cache.auth.memcache_secret_key < "${DEST_DIR}/secret-openstack.yaml"')
```

File: scripts/gitops-secrets-gen.sh [hunk 5]
```diff
@@ -413,20 +148,240 @@ spec:
...
+export LC_ALL=C
+## OpenStack component secret generation
+## each openstack component is very similar to collapse this
+## into a loop to generate the same thing for each
+for component in keystone ironic placement neutron nova glance; do
+    echo "Checking ${component}"
+    mkdir -p "${DEST_DIR}/${component}/"
+    # keystone service account username
+    [ "x${component}" = "xkeystone" ] && keystone_user="admin" || keystone_user="${component}"
+    KEYSTONE_USER=$(echo "${keystone_user}" | tr '[:lower:]' '[:upper:]')
```
Comment (skrobul) [inline scripts/gitops-secrets-gen.sh:256]: is this used anywhere?

File: scripts/gitops-secrets-gen.sh [hunk 6]
```diff
@@ -413,20 +148,240 @@ spec:
...
+        REPLACE_KEYSTONE_PASSWORD=no
+    else
+        echo "Generating ${SECRET_KEYSTONE_PASSWORD}"
+        declare ${KEYSTONE_PASSWORD}="$(./scripts/pwgen.sh 2>/dev/null)"
+        REPLACE_KEYSTONE_PASSWORD=yes
+    fi
+    # export the variables for templating the openstack secret
+    export ${RABBITMQ_PASSWORD}
+    export ${DB_PASSWORD}
+    export ${KEYSTONE_PASSWORD}
```
Comment (skrobul) [inline scripts/gitops-secrets-gen.sh:297]:
```suggestion
    export ${RABBITMQ_PASSWORD?}
    export ${DB_PASSWORD?}
    export ${KEYSTONE_PASSWORD?}
```

**Review by skrobul — COMMENTED** [review]
good stuff, it already looks better. could use some DRY in a few places as per the comments.

File: scripts/gitops-secrets-gen.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (cardoe) [reply scripts/gitops-secrets-gen.sh:261]: Can I go with `VARNAME`? Feel like that's more obvious?

[the remark this answers is not in the document]
Comment (cardoe) [reply scripts/gitops-secrets-gen.sh:261]: Marek said yes in chat.

Comment (cardoe) [reply scripts/gitops-secrets-gen.sh]: Fighting wish bash 3 locally. :(

File: scripts/gitops-secrets-gen.sh (hunk 5 above)
Comment (cardoe) [reply scripts/gitops-secrets-gen.sh]: Nope. Deleted

File: scripts/gitops-secrets-gen.sh (hunk 6 above)
Comment (cardoe) [reply scripts/gitops-secrets-gen.sh]: Done.

**Comment by cardoe** [comment]
Rebased on #226

**Review by skrobul — APPROVED** [review]

**Pull request merged**